---
layout: default
title: "Walkthrough"
date: "2025-01-11"
permalink: /walkthrough/

---
# In Class Walkthrough 

![image](https://github.com/user-attachments/assets/26279a58-b4ba-4dfc-8691-84ff4936b375)

- The "hands-on" portion of the class will focus on trying to manipulate 16S abundance tables so that we can pull out informative data and visualizations 

- We won't have time to go through the entirety of the 16S analysis pipeline. I've pre-processed the raw 16S reads to do the following: 1) Import raw 16S reads into Qiime2, 2) Trim primers and adapters, 3) denoising data (forming ASVs), 4) generating a feature abundance table, 5) assigning ASV taxonomy

- If you are interested in how to do this, the code is available in the "data pre-processing" section

- The rest of the analyses that we will be going through together will be in R. Don't worry if you don't have much R experience. Most of the code below can be directly copied and pasted. If that is still an issue and/or package installation proves to be challenging, you can pair up with someone else in the class. The main goal here is not necessarily to learn the exact R syntax, but rather to understand the analysis workflow and to get a grasp of the types of information we can and cannot get from 16S sequencing data 


## Importing the data 

- We will exclusively be working in R, but before we do so it's important to understand the data types we will be working with and importing

- I used Qiime2 (see resources) for pre-processing, which involves storing data in file types called "artifacts" in .qza files. We can import the artifacts into R and convert them into a phyloseq object (this is will make more sense below)

- Phyloseq is an R package that integrates: 1) abundance tables (ie- the abundance of each ASV in each sample), 2) taxonomy information (ie - what is the taxonomic identity of each ASV?), 3) metadata (what are features of the samples themselves?), 4) phylogeny information (what is the phylogenetic relatedness of the ASVs in the dataset?)

- These 4 features can be integrated into a single phyloseq "object" in R. Having all this information integrated into a single object makes our code cleaner and more straightforward to run :) 


- Please download the following artifacts and files into the same directory: 1) abundance artifact, 2) taxonomy artifact, 3) metadata.txt, 4) phylogeny artifact


### Installing R packages 

- Our first step will be integrating these artifacts into a phyloseq object

- Before we do that, we'll have to install some R packages. These will include some basic R libraries for data manipulation and visualization (dplyr and ggplot2), phyloseq, and packages that use phyloseq (qiime2R, microViz, ape, microbiome)

- Below are instructions to install these libraries. If you can, try to install these before class, but we can spend a bit of time ensuring we are all setup properly


```{r}

###Install phyloseq###

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")


###Install ape###

install.packages("ape")


###Install qiime2R###

if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
devtools::install_github("jbisanz/qiime2R")


###Install Microviz###

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("phyloseq", "microbiome", "ComplexHeatmap"), update = FALSE)



###Install Microbiome###

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("microbiome")


```


- Now load the installed packages!

```{r}
library(phyloseq)
library(qiime2R)
library(microViz)
library(ggplot2)
library(ape)
library(microbiome)
library(dplyr)
```


- Now that our packages are installed and loaded, we can import our Qiime2 artifacts and create a phyloseq object in one simple step using qza_to_phyloseq
- Simply replace the path file in quotations with the path to the file in your directory (right-click, option, copy path-name on mac)


```{r}
physeq <- qza_to_phyloseq(
    features="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/feature-sample-filtered-table.qza",
    tree="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/rooted-tree.qza",
    taxonomy="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/filtered-taxonomy.qza",
    metadata = "/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/sample-metadata.txt"
    )
```


## Section 1: Exploring the Data

-  We've now succesfully imported our data into a phyloseq object!
-  View the phyloseq object to get an idea of the data we are working with?

```{r}
physeq
```


-  **Question 1A) How many taxa (ASVs) and samples do we have total in the dataset?**


- Now lets explore the metadata a bit to get a sense of the dataset-specific parameters we're dealing with
- We can access different features of the phyloseq object
- To acess the metadata specifically we can do the following:

```{r}
sample_data(physeq)

#and save it as a dataframe 
metadata <- as.data.frame(sample_data(physeq))
```

```{r}
#You can use unique() in base R to find all unique values in a specific column 
#For example
unique(metadata$Day)

```

 **Question 1B) How many time points are in the experiment?**
 **Question 1C) What are the three diet treatments the mice are given?**
 **Question 1D) Where are the mouse vendors from?**
 
Start to think of some microbiome based questions one could answer based on the metdata we have!


## Section 2: Exploring Microbiome Composition

- Now that we have a sense of the samples we are working with, lets try to visualize some data
- A good way to get a first overview of your microbiomes are with simple bar plots. These aren't perfect, but they can give us a nice first impression of what to expect
- We can a simple bar plot using the Microviz comp_barplot command 
- This command tells will plot - per sample - the relative abundance of bacteria at different taxonomic levels 
- Lets first look at some high-level taxonomy. 
- First we need to "fix" the names taxa - removes possible unclassified ("NAs") from our dataset, which are incompatible with the comp_barplot function

```{r}
fixed_physeq <- tax_fix(physeq)
fixed_physeq <- fixed_physeq %>% 
  tax_fix(unknowns = c("uncultured"))
```

- Our "fixed" phyoloseq can now be used to generate relative abundance plots downstream 
- Now plot using the comp_barplot function - use the "fixed" phyloseq object and toggle between different taxonomic ranks using the "tax_level" flag 
- Spend some time looking through these plots at different taxonomic levels 

```{r}
comp_barplot(fixed_physeq, tax_level = "Phylum") 
```

![image](https://github.com/user-attachments/assets/b7a2bbe8-2535-46f0-9305-727b4be019c9)


-  Note that by default, comp_barplot will plot the most 8 abundant taxa, you can increase this number by using the n_taxa flag - more useful for lower level ranks 

```{r}
comp_barplot(fixed_physeq, tax_level = "Phylum", n_taxa = 12) 
```

Does anything stand out here? 
**Question 3A) Which bacterial phyla tend to be the most abundant in the dataset?**
**Question 3B) What about at the family and genus levels?**
**Question 3C) Try doing this at species level, anything strange to report of the species names?**


## Part 3 - Exploring ASV trends 

- We have nice broad view of  high level taxonomy in the data, but what about at finer resolution? 
- Since we have a large number of distinct ASVs in our dataset, it's hard to visualize the data using something like a bar plot or even a heatmap
- We can see how prevalent each of these ASVs are across the dataset to get a sense of which ones are "important"
- This will require a bit of data manipulation... The code below transforms our phyloseq object to a dataframe which gives the number of samples that each ASV is found in


```{r}
#For folks with R experience: we are applying a function that 1) accesses the OTU table from the Phyloseq, 2) MARGIN =1 defines that our taxa are in the rows, 3) FUN: for each row (ASV) it counts the number of non-zero entries 


prevalence = apply(X = otu_table(physeq),MARGIN = 1, FUN = function(x){sum(x > 0)})


#convert to a df
prevalence.df <- as.data.frame(prevalence)

# Our ASVs (rows) have indecipherable codes (used by the Silva database)!
# Lets use our taxonomy table to help us find out the ID of these 




# 1) Get the taxonomy table from the phyloseq object! 
tax_df <- as.data.frame(tax_table(physeq))


#2) now merge the taxonomy dataframe and the prevalence dataframe 
merged_df_prevalence <- merge(prevalence.df, tax_df, by = "row.names", all = TRUE)


```

- Now we have a dataframe that has each unique ASV as a row name, taxonomy labels, and how many samples each ASV is found in.

- Before we think about ASV prevalence, take a look at the dataframe and look at the taxonomy annotations. Notice that we have NO species-level assignments. This is not a mistake. In general, V3-V4 primers don't give great species resolution. The Silva database also doesn't curate species names in their database, so for accuracy sake we wont consider these. 


- You can use sum(!is.na(df$your_column)) to count the number of rows that are NOT NA (ie- have a taxonomy assignment). Ie - 0 for species


```{r}
sum(!is.na(merged_df_prevalence$Species))
```

**Question 3A) How many ASVs have genera and family-level designations?**



- Now we can dive into some of the prevalence data a bit more
- Order the order the dataframe in descending order to get an idea of the most prevalent ASVs


```{r}
df_prevalence_ordered <- merged_df_prevalence[order(-merged_df_prevalence$prevalence), ]
```

**Question 3B) How many ASVs are found in ALL 152 samples and what genera do they belong to?**


- Let's do something similar but with abundance instead of prevalence!

```{r}
# Step 1: Transform the phyloseq object to get relative abundance info for each ASV using the "transform_sample_counts" function
#Now we have a phyloseq object that has relative abundance counts instead of "raw counts"

physeq_rel <- transform_sample_counts(physeq, function(x) x / sum(x))



# Step 2: Extract abundance data from our phyloseq object 

taxa_abundance <- taxa_sums(physeq_rel)  # Sum the relative abundance for each ASV

#Turn into a dataframe
abundance_df <- data.frame(ASV = names(taxa_abundance), Abundance = taxa_abundance)



# Step 3: Merge with our taxonomy dataframe
merged_df_abundance <- merge(abundance_df, tax_df, by = "row.names", all = TRUE)

#Now we have 


# Step 4: Rank taxa by abundance
#ordering the dataframe by abundance 

merged_df_abundance <- merged_df_abundance[order(-merged_df_abundance$Abundance), ]
#creating a rank using seq_len, which gives a integer in order to the ordered row names
merged_df_abundance$Rank <- seq_len(nrow(merged_df_abundance))


```

- Now we can view our abundance dataframe, which ranks each ASV by its abundance across samples in the dataset


**Question 3C) What are the 5 most abundant ASVs in the dataset?**
**Question 3D) Are these also the most prevalent ones? (Compare your abundance dataframe with your prevalence dataframe)**


- Let's make a Whittaker plot. These are rank-abundance curves used in ecology. Can give us an idea of the ASVs that are rare and dominant

```{r}
ggplot(merged_df_abundance, aes(x = Rank, y = Abundance)) +
  geom_line() +
  geom_point() +
  labs(
    title = "Rank-Abundance Curve",
    x = "Rank",
    y = "Relative Abundance"
  ) +
  theme_minimal()
```

![image](https://github.com/user-attachments/assets/576683ac-5b90-4262-83ce-ded06f47308c)


- Note the long tail, which indicates that there are a small number of highly abundant ASVs and a long-tail of low-abundant "rare" ASVs
- A common feature of gut microbiomes!

## Part 4 - Explore dataset-specific trends using bar plots

- So far we've explored the metadata and now have a nice overview of what taxa are present in the dataset
- Now we can try to look for trends to gain some insight into the study
- A reminder that this study deals looks at the effect of diet intervention in mice orginating from different vendors
- Lets see if the vendor the mice come from has any effect on composition using the comp_barplot function that we used earlier. 
- Now we can add the group_by flag to group our plots by a metadata column!


```{r}
comp_barplot(fixed_physeq, tax_level = "Family", group_by = "vendor", n_taxa = 12)
```

**Question 4A) Look through different taxonomic levels - What differences in the relative abundance do you observe based on mouse vendor?**
**Question 4B) In the published paper they describe differences specifically in Muribaculaceae (polysaccharide-degrading). Describe the differences in this family between vendors**



- It does look like the vendor influences microbiome composition! To determine the effect of diet on these mice we may have to account for this 
- One strategy is to analyze the effects of diet separately for each vendor. To do this we'll "filter" our phyloseq object for each vendor 
- We can use the subset_samples function to 
- To make our life easier subset from the "fixed" ps object 

```{r}
bejing_physeq <- subset_samples(fixed_physeq, vendor == "Beijing")
# view the physeq object to confirm that we're filtering samples as expected!


# Do this for each vendor 
shanghai_physeq <- subset_samples(fixed_physeq, vendor == "Shanghai")

guangdong_physeq <- subset_samples(fixed_physeq, vendor == "Guangdong")
```


- Now we can start to look at the effect of diet on the microbiomes of these mice...
- But first there is one more consideration. Time point! 
- 0 = baseline (before diet intervention), and 5, 13, 31 says post intervention (Note that time points vary between vendor study)
=  For each vendor, we can investigate the effect of diet at different time points by faceting our plot!


```{r}

#For simplicity sake, let's just focus on Guangdong and Shanghai at the Family level

comp_barplot(shanghai_physeq, tax_level = "Family", group_by = "Group", facet_by = "Day", n_taxa = 12)
comp_barplot(guangdong_physeq, tax_level = "Family", group_by = "Group", facet_by = "Day", n_taxa = 12)

```


**Question 4C) In these studies cellulose is the control diet (all mice are kept on cellulose before day 0, and then either switch to inulin, resistant starch or are kept on cellulose). Is microbiome composition stable over time in this control?**

**Question 4D) What effects do inulin and resistant starch have on the microbiome? Are these constant over time?  do they differ based on mouse vendor?**

